### PR TITLE
Rebase #PR-101 against master

### PR DESCRIPTION
**Pull request by cdancy**

### CODE REVIEW

**Comment by cdancy** [comment]
@bmuschko PR for rebasing #101 against master. Wanted to get a final check in as last commit was from September. Everything looks good. Any objections?

File: src/test/groovy/com/bmuschko/gradle/docker/utils/TaskStateHelperTest.groovy [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.bmuschko.gradle.docker.utils
+
+import org.gradle.api.Project
+import spock.lang.Specification
+
+class TaskStateHelperTest extends Specification {
```
Comment (bmuschko) [inline src/test/groovy/com/bmuschko/gradle/docker/utils/TaskStateHelperTest.groovy]: I'd probably move this into the integration test SourceSet and use `ProjectBuilder` instead.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 2]
```diff
@@ -18,8 +18,10 @@ package com.bmuschko.gradle.docker.tasks.image
 import com.bmuschko.gradle.docker.DockerRegistryCredentials
 import com.bmuschko.gradle.docker.tasks.AbstractDockerRemoteApiTask
 import com.bmuschko.gradle.docker.tasks.RegistryCredentialsAware
+import com.bmuschko.gradle.docker.utils.TaskStateHelper
 import org.gradle.api.tasks.*
 
+
 class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCredentialsAware {
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: Instead of shoving this logic into this class I'd create a new class and delegate from there. That'll separate logic, make it more readable and maintainable.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/AbstractDockerRemoteApiTask.groovy [hunk 3]
```diff
@@ -44,11 +43,23 @@ abstract class AbstractDockerRemoteApiTask extends DefaultTask {
...
     @TaskAction
     void start() {
-        threadContextClassLoader.withClasspath(getClasspath().files, createDockerClientConfig()) { dockerClient ->
+        runInDockerClassPath { dockerClient ->
             runRemoteCommand(dockerClient)
         }
     }
 
+    @InputFiles
+    public File[] getCertPathKeys() {
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/AbstractDockerRemoteApiTask.groovy]: This doesn't look related to the actual purpose of the change. I'd remove this for now and deal with it in a separate PR.

**Comment by bmuschko** [comment]
We need to have a functional test that demonstrates the it works properly.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 4]
```diff
@@ -117,4 +122,38 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
         imageId = response.awaitImageId()
         logger.quiet "Created image with ID '$imageId'."
     }
-}
\ No newline at end of file
+
+    void saveImageId() {
+        def state = new TaskStateHelper(name, project)
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: Let use concrete types.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 5]
```diff
@@ -117,4 +122,38 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
...
+        state.put("imageId", getImageId())
+    }
+
+    boolean previouslyBuiltImageExists() {
+        def imageExists = false
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: Let use concrete types.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 6]
```diff
@@ -117,4 +122,38 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
...
+
+        def prevImageId = new TaskStateHelper(name, project).get("imageId")
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: Let use concrete types.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 7]
```diff
@@ -117,4 +122,38 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
...
+                logger.info "Image ${prevImageId} found via call to inspectImage"
+
+                imageExists = true
+            } catch(Exception e) {
+                logger.info "Image ${prevImageId} not found via call to inspectImage"
+                imageExists = false
+            }
+        }
+
+        if (imageExists) {
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: I think we can just move this logic into the `try` clause and return from there.

File: src/main/groovy/com/bmuschko/gradle/docker/utils/TaskStateHelper.groovy [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+import org.gradle.api.Project
+
+class TaskStateHelper {
+
+    String taskName
+
+    Project project
+
+    public TaskStateHelper(String taskName, Project project) {
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/utils/TaskStateHelper.groovy]: Instead of passing in the Project instance, let's just pass in `File buildDir`. That'll decouple the Gradle API from this class and allows us to write unit tests. Let's also write a unit test or turn the existing on into a unit test.

**Comment by cdancy** [comment]
@bmuschko take a peek when you get a chance. Addressed your concerns and simplified TaskStateHelper. Thanks.

**Comment by bmuschko** [comment]
@cdancy Can we apply PRs separately if possible? I think you also incorporated #100. Let's remove it. Also I'd like to keep the UP-TO-DATE check.

**Comment by cdancy** [comment]
Commit has been removed. @bmuschko I'm not exactly sure what you mean by "Also I'd like to keep the UP-TO-DATE check." I did refactor the test to check for "UP-TO-DATE" message if that's what you mean...

**Comment by bmuschko** [comment]
@cdancy I meant this here: https://github.com/bmuschko/gradle-docker-plugin/pull/139/files#diff-5cc5cb1ca53ae2ec53c1ac78beba9dd0R76

**Comment by cdancy** [comment]
@bmuschko Ok. That snippet is indeed present with the latest commit so I believe we should be good on that end.

**Comment by cdancy** [comment]
Merged to master. @ajmath thanks for the commit!

**Comment by ajmath** [comment]
Awesome!  Thanks for seeing it through
